### PR TITLE
Add cudnn v4 batch normalization integration

### PR DESCRIPTION
**Pull request by andrei-pokrovsky**
Tested on a simple cifar10-quick modified network with one extra added layer to produce about the same loss/accuracy as caffe's implementation of BN.

Example of a BN layer inserted after relu1 into cifar10_quick:
```
layer {
  name: "BN1"
  type: "BatchNorm"
  bottom: "pool1"
  top: "bn1"

  batch_norm_param {
    scale_filler {
      type: "constant"
      value: 1
    }
    bias_filler {
      type: "constant"
      value: 0.001
    }
    engine: CUDNN
  }
}
```
Note that to switch to Caffe's implementation of BN the option is engine: CAFFE

(Previous layer's name is pool1 because relu1 is in-place)

### CODE REVIEW

File: src/caffe/layers/batch_norm_layer.cpp [hunk 1]
```diff
@@ -232,5 +232,4 @@ STUB_GPU(BatchNormLayer);
 #endif
 
 INSTANTIATE_CLASS(BatchNormLayer);
-REGISTER_LAYER_CLASS(BatchNorm);
```
Comment (thatguymike) [inline src/caffe/layers/batch_norm_layer.cpp]: Why this change?

Comment (thatguymike) [inline src/caffe/layers/batch_norm_layer.cpp]: Ah, because of the change in registration to switch between CAFFE and CUDNN engines, right?

**Comment by thatguymike** [comment]
[  FAILED  ] CuDNNBatchNormLayerTest/0.TestForward, where TypeParam = float
[  FAILED  ] CuDNNBatchNormLayerTest/1.TestForward, where TypeParam = double

Andrei - can you take a look?

**Comment by thatguymike** [comment]
Specifically:

channel: 0, mean: 0, var: 0
src/caffe/test/test_batch_norm_layer.cpp:184: Failure
The difference between 1 and var is 1, which exceeds kErrorBound, where
1 evaluates to 1,
var evaluates to 0, and
kErrorBound evaluates to 0.001.

File: src/caffe/test/test_batch_norm_layer.cpp [hunk 2]
```diff
@@ -130,4 +130,82 @@ namespace caffe {
...
+
+  Dtype mean, var;
+  for (int j = 0; j < channels; ++j) {
+    mean = 0;
+    var = 0;
+    for (int i = 0; i < num; ++i) {
+      for (int k = 0; k < height; ++k) {
+        for (int l = 0; l < width; ++l) {
+          Dtype data = top->data_at(i, j, k, l);
+          printf("data: %lg\n", static_cast<double>(data));
```
Comment (thatguymike) [inline src/caffe/test/test_batch_norm_layer.cpp]: Probably remove the printf

File: src/caffe/test/test_batch_norm_layer.cpp [hunk 3]
```diff
@@ -130,4 +130,82 @@ namespace caffe {
...
+          mean += data;
+          var += data * data;
+        }
+      }
+    }
+    mean /= num * height * width;
+    var /= num * height * width;
+
+    printf("channel: %d, mean: %lg, var: %lg\n", j, mean, var);
```
Comment (thatguymike) [inline src/caffe/test/test_batch_norm_layer.cpp]: Remove printf or move to info log.

File: src/caffe/layers/batch_norm_layer.cpp (hunk 1 above)
Comment (andrei-pokrovsky) [inline src/caffe/layers/batch_norm_layer.cpp]: Correct.

**Comment by thatguymike** [comment]
Merged.  Closes #80
